text/000-buildpacks-contribute-default-process-type.md

```diff
@@ -0,0 +1,113 @@
+
+- Why should we do this? Easier and more intuitive for end users.
+- What use cases does it support? Apps with one or more process type(s).
+- What is the expected outcome? The experience should be seamless for end users. Buildpacks should produce a sensible default process type without the user having to know or care that process types exist. The user should still be able to override the default process type by passing `-process-type` to the lifecycle.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Example: single process app with a buildpack declaring a `default-process` of type `web` -> the app image will have a default process type of `web`
+
+Example single process app with no buildpacks declaring a `default-process` -> the app image will have no default process. Users must specify the process to run at runtime. (This assumes https://github.com/buildpacks/spec/pull/137)
+
+Example: multi-process app with a buildpack declaring a `default-process` of type `web` -> the app image will have a default process type of `web`
+
+Example: multi-process app with an earlier buildpack declaring a `default-process` of type `web` and a later buildpack declaring a process of type `worker` -> the app image will have a default process type of `worker`
+
+Example: multi-process app with a buildpack declaring a `default-process` of type `web` and the user passes `-process-type worker` -> the app image will have a default process type of `worker`
+
+Example: single process app with no buildpacks declaring a `default-process` and the user passes `-process-type worker` -> the app image will have a default process type of `worker`
+
+- Define the target persona: buildpack author, buildpack user.
+
+New buildpack users would need to be taught that buildpacks will be responsible for setting a sensible default process type. They should know how to override the buildpacks-provided default if they desire.
+
+Existing buildpack users would need to be taught about the change so that (1) they no longer have to pass a flag to specify the default process type or (2) they now need to pass a flag to explicitly override the default that buildpacks choose. For (2), it's likely that the number of users in this category would be small given that the lifecycle (pre- platform API 0.4) and pack (post- platform API 0.4) both encourage `web` as the default, and it's likely that buildpacks would also encourage this same default.
+
+Buildpack authors would need to be taught how to configure the default process type.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The changes from this RFC should require a bump to the buildpack API.
+
+Currently, during the `build` phase, buildpacks may declare process types by writing entries to `<layers>/launch.toml`. This RFC proposes a new key in `<layers>/launch.toml`, `[default-process]`, that specifies what each buildpack would like to designate as the default process type for the app image.
+
+```toml
+[default-process]
```

Comment:
I wonder if we should consider adding a boolean `override` key to the `default-process` table. Then buildpacks could indicate either "this should be the default if there is none" (`override = false`) or "make this the default regardless" ( `override = true`). This would add slightly more complexity but might be very nice to have in certain use cases.

If we added this key I suspect `override` should be `true` by default.

Reply:
Ah. So if a buildpack with `override = true` sets the default process, the user cannot override?

Reply:
@ekcasey Can you give an example of where this might be useful?  It seems like this effectively inverts the precedence chain (first wins instead of last wins) and could be achieved by a user doing the same but changing the order of the buildpacks.

Reply:
@natalieparellano No, I don't think it stops a user from contributing.  Envision this as all the buildpacks do their default process type contributions and if the user has specified one, theirs comes last and would override everything else.

This change is a suggestion such that right now a chain of A -> B -> C always results in C being the selected default where as `A:false -> B:true -> C:false` results in B being the default.  The flag is whether _that_ default will override one prior to it.

Reply:
Added the concept of `override` to the RFC.

Reply:
Removed the concept of `override` per most recent discussion in WG.